File: sinatra-ruby/app/controllers/api/sca/v1/actions_controller.rb
#
# @author Daniel Marcenco (danielm@saltedge.com)
# Copyright (c) 2022 Salt Edge.
#

class ActionsController < ApiBaseController
  namespace '/api/sca/v1/actions' do
    post '/:action_id' do
      action = Action.find_by(id: params['action_id'])

      raise ActionNotFound unless action

      connection_id = params['data']['connection_id']

      raise BadRequest::WrongRequestFormat unless connection_id

      connection = Connection.find_by(id: connection_id)

      raise ConnectionNotFound unless connection

      create_action_payload(action, connection)
    end

    put '/:action_id/confirm' do
      action = Action.find_by(code: params['action_id'])

      raise ActionNotFound unless action
      raise ActionClosed if action.closed?

      ConfirmAuthorizationsWorker.perform_async(
        action_id: action.id,
        authorization_code: decrypted_request['authorization_code'],
        is_confirm: true
      )

      {
        data: {
          close_action: true
        }
      }.to_json
    end

    put '/:action_id/deny' do
      action = Action.find_by(code: params['action_id'])

      raise ActionNotFound unless action
      raise ActionClosed if action.closed?

      ConfirmAuthorizationsWorker.perform_async(
        action_id: action.id,
        authorization_code: decrypted_request['authorization_code'],
        is_confirm: false
      )

      {
        data: {
          close_action: true
        }
      }.to_json
    end

    private

    def create_action_payload(action, connection)
      authorization           = EncryptionTools.create_authorization_data(action)
      encrypted_authorization = EncryptionTools.create_encrypted_action(authorization, connection)

      {
        data: {
          action_id: action.id.to_s,
          user_id: SCA_USER_ID,
          multiple: action.multiple,
          expires_at: action.expires_at.to_s,
          authorization: encrypted_authorization
        }
      }.to_json
    end

    def parsed_original_request
      JSON.parse(params['data']['original_request'])['data']
    end

    def decrypted_request
      JSON.parse(Jws.decrypt_public_rsa_key(parsed_original_request))
    end
  end
end

File: spring-boot/src/main/java/com/saltedge/provider/demo/controllers/api/sca/v1/ActionsController.kt
/*
 * @author Constantin Chelban (constantink@saltedge.com)
 * Copyright (c) 2022 Salt Edge.
 */
package com.saltedge.provider.demo.controllers.api.sca.v1

import com.saltedge.provider.demo.config.SCA_USER_ID
import com.saltedge.provider.demo.controllers.api.sca.v1.model.*
import com.saltedge.provider.demo.errors.BadRequest
import com.saltedge.provider.demo.errors.NotFound
import com.saltedge.provider.demo.model.ScaActionsRepository
import com.saltedge.provider.demo.model.ScaConnectionsRepository
import com.saltedge.provider.demo.tools.createAuthorizationData
import com.saltedge.provider.demo.tools.createEncryptedAction
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping(ActionsController.BASE_PATH)
class ActionsController : BaseController() {
    companion object {
        const val BASE_PATH: String = "$API_BASE_PATH/actions"
    }

    @Autowired
    lateinit var actionsRepository: ScaActionsRepository
    @Autowired
    lateinit var connectionsRepository: ScaConnectionsRepository

    @Throws(Exception::class)
    @PostMapping("/{action_id}")
    fun createAuthorizationInfo(
        @PathVariable("action_id") actionId: Long,
        @RequestBody request: CreateAuthorizationInfoRequest
    ): ResponseEntity<CreateAuthorizationInfoResponse> {
        val action = actionsRepository.findById(actionId).orElse(null) ?: throw NotFound.ActionNotFound()
        val connectionIdValue = request.data.connectionId ?: throw BadRequest.WrongRequestFormat()
        val connection = connectionsRepository.findFirstByConnectionId(connectionIdValue)    ?: throw NotFound.ConnectionNotFound()

        val authorization = createAuthorizationData(action)
        val encAuthorization = createEncryptedAction(data = authorization, connection = connection)

        return ResponseEntity(CreateAuthorizationInfoResponse(data = CreateAuthorizationInfoResponseData(
            action_id = actionId.toString(),
            user_id = SCA_USER_ID,
            expires_at = action.expiresAt.toString(),
            authorization = encAuthorization
        )), HttpStatus.OK)
    }

    @Throws(Exception::class)
    @PutMapping("/{action_id}/confirm")
    fun confirm(
        @PathVariable("action_id") actionId: Long,
        @RequestBody request: UpdateActionRequest
    ): ResponseEntity<UpdateActionResponse> {
        val action = actionsRepository.findById(actionId).orElse(null) ?: throw NotFound.ActionNotFound()

        if (action.status == "confirmed" || action.status == "denied") throw BadRequest.ActionClosed()

        action.status = "confirmed"
        actionsRepository.save(action)
        return ResponseEntity(UpdateActionResponse(data = UpdateActionResponseData(close_action = true)), HttpStatus.OK)
    }

    @PutMapping("/{action_id}/deny")
    fun deny(
        @PathVariable("action_id") actionId: Long,
        @RequestBody request: UpdateActionRequest
    ): ResponseEntity<UpdateActionResponse> {
        val action = actionsRepository.findById(actionId).orElse(null) ?: throw NotFound.ActionNotFound()

        if (action.status == "confirmed" || action.status == "denied") throw BadRequest.ActionClosed()

        action.status = "denied"
        actionsRepository.save(action)
        return ResponseEntity(UpdateActionResponse(data = UpdateActionResponseData(close_action = true)), HttpStatus.OK)
    }
}